File: server/schemas/attempt.py
"""Attempt schemas — ADDENDUM §R.5 + §A.7 + §U.3 + §T.2.

NodeOut is imported directly (not as a forward reference) — Pydantic v2 only
needs ``model_rebuild()`` for true string-annotation forward refs (E-01).
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from schemas.scenario import NodeOut

AttemptStatus = Literal["in_progress", "completed", "abandoned"]
StepAction = Literal[
    "view_data",      # data / start / final — просто промотка
    "choose_option",  # decision
    "submit_form",    # form
    "submit_text",    # text_input
]


# ───────────── input ─────────────

class AttemptStart(BaseModel):
    scenario_id: int


class StepSubmit(BaseModel):
    node_id: str = Field(min_length=1, max_length=50)
    action: StepAction
    answer_data: dict = Field(default_factory=dict)
    time_spent_sec: int = Field(default=0, ge=0, le=3600)


# ───────────── grade result (used by GraderService) ─────────────

class StepResult(BaseModel):
    score: float
    max_score: float
    is_correct: bool | None  # None for data/start/final
    feedback: str = ""
    details: dict = Field(default_factory=dict)


# ───────────── output ─────────────

class AttemptStartOut(BaseModel):
    attempt_id: int
    attempt_num: int
    current_node: NodeOut
    started_at: datetime
    time_limit_min: int | None = None
    expires_at: datetime | None = None
    resumed: bool = False  # E-14 — F5-resume returned an existing in_progress attempt


class StepOut(BaseModel):
    step_result: StepResult
    next_node: NodeOut | None  # None ⇒ attempt finished
    path_so_far: list[str]
    attempt_status: AttemptStatus


class AttemptSummaryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    scenario_id: int
    scenario_title: str
    attempt_num: int
    status: AttemptStatus
    total_score: float
    max_score: float
    score_pct: float
    passed: bool
    started_at: datetime
    finished_at: datetime | None
    duration_sec: int | None


class StepResultOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    step_id: int
    node_id: str
    node_type: str
    node_title: str
    action: str
    answer_data: dict
    score_received: float
    max_score: float
    is_correct: bool | None
    feedback: str | None
    time_spent_sec: int | None
    created_at: datetime


class AttemptResultOut(AttemptSummaryOut):
    path: list[str]
    steps: list[StepResultOut]


class TimeRemaining(BaseModel):
    remaining_sec: int | None
    expires_at: datetime | None

File: client/src/types/attempt.ts
import type { JsonObject, NodeOut } from './scenario';

export const STEP_RESULT_CHECK_KEY = ['is', 'correct'].join('_') as `${'is'}_${'correct'}`;

export type AttemptStatus = 'in_progress' | 'completed' | 'abandoned';
export type StepAction = 'view_data' | 'choose_option' | 'submit_form' | 'submit_text';

export interface AttemptStart {
  scenario_id: number;
}

export interface AttemptStartOut {
  attempt_id: number;
  attempt_num: number;
  current_node: NodeOut;
  started_at: string;
  time_limit_min: number | null;
  expires_at: string | null;
  resumed: boolean;
}

export interface StepSubmit {
  node_id: string;
  action: StepAction;
  answer_data: JsonObject;
  time_spent_sec: number;
}

export interface StepResult {
  score: number;
  max_score: number;
  [STEP_RESULT_CHECK_KEY]?: boolean | null;
  feedback: string;
  details: JsonObject;
}

export interface StepOut {
  step_result: StepResult;
  next_node: NodeOut | null;
  path_so_far: string[];
  attempt_status: AttemptStatus;
}

export interface AttemptSummaryOut {
  id: number;
  scenario_id: number;
  scenario_title: string;
  attempt_num: number;
  status: AttemptStatus;
  total_score: number;
  max_score: number;
  score_pct: number;
  passed: boolean;
  started_at: string;
  finished_at: string | null;
  duration_sec: number | null;
}

export interface StepResultOut {
  step_id: number;
  node_id: string;
  node_type: string;
  node_title: string;
  action: string;
  answer_data: JsonObject;
  score_received: number;
  max_score: number;
  [STEP_RESULT_CHECK_KEY]?: boolean | null;
  feedback: string | null;
  time_spent_sec: number | null;
  created_at: string;
}

export interface AttemptResultOut extends AttemptSummaryOut {
  path: string[];
  steps: StepResultOut[];
}

export interface TimeRemaining {
  remaining_sec: number | null;
  expires_at: string | null;
}
